### ISSUE TITLE
Installing a version of plugin sometimes fails on Windows

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi#2695]
```
pulumi plugin install resource azure v0.18.2-dev.1556768307+g78c053
[resource plugin azure-0.18.2-dev.1556768307+g78c0537] installing
Downloading plugin:  36.55 MiB / 36.55 MiB [=======================] 100.00% 22s
error: installing [resource plugin azure-0.18.2-dev.1556768307+g78c0537] from https://api.pulumi.com:
moving plugin: rename C:\Users\name\.pulumi\plugins\resource-azure-v0.18.2-dev.1556768307+g78c0537.tmp804464559 C:\Users\name\.pulumi\plugins\resource-azure-v0.18.2-dev.1556768307+g78c0537: 
Access is denied.
```

The temporary folder does get created and the manual rename does work, so I suspect some API which fails for this kind of path?

## Edit: Summary of the underlying issue and proposed solution

This exact same problem was happening with golang's module cache on Windows: module zips were downloaded/extracted to a temp dir and then renamed with `os.Rename` to the final dir, just like how we handle plugin installs (https://github.com/golang/go/issues/36568). The problem on Windows is that aggressive file scanners are checking the files in the temp dir, preventing the rename operation from succeeding (errors with "Access is denied.").

The golang team's solution was to extract the zip in-place in the final destination, rather than extracting to a temp dir and renaming it. Note that they were leveraging some file locking mechanism they already had in place, and we'll likely need to do something similar to prevent multiple concurrent installs from cobbling over each other.

A simpler initial "fix" may be just to introduce some retries. If I'm remembering correctly, the golang team initially tried this, but still ran into the problem. However, the amount of plugin files is typically very small (~1-2 files), so we may have better luck with retries since we're not dealing with that many files. Though, as more multi-lang components are introduced, the number of files in plugin directories will also increase as those will be running `npm install` (and equivalents for the other languages).

Reference:
https://github.com/golang/go/commit/093049b3709eda7537ece92a2991918cf53782d6
https://github.com/golang/go/commit/576fa692774137633b09dd244e1de36993dd2803

**Comment by lukehoban** [comment]
Does this only happen for dev versions?  That is particularly odd.

**Comment by mikhailshilkov** [comment]
This week is the first time I saw this issue, and I already got it twice for two dev versions...

**Comment by mikhailshilkov** [comment]
Okay, it's not just dev versions. Got the same issue for aws 0.18.5 and azure 0.18.4 today.

**Comment by mikhailshilkov** [comment]
A friend of mine got a similar error while trying to get started with Pulumi+AWS+python on Windows:

```
error: installing [resource plugin aws-0.18.11] from : moving plugin: rename C:\Users\User\.pulumi\plugins\resource-aws-v0.18.11.tmp448077619 C:\Users\User\.pulumi\plugins\resourc
e-aws-v0.18.11: Access is denied.
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "C:\Users\user\AppData\Local\Temp\pip-install-p9rgnm5n\pulumi-aws\setup.py", line 36, in <module>
        zip_safe=False)
      File "c:\users\user\dropbox\pocs\quickstart\venv\lib\site-packages\setuptools\__init__.py", line 145, in setup
        return distutils.core.setup(**attrs)
      File "c:\python36\Lib\distutils\core.py", line 148, in setup
        dist.run_commands()
      File "c:\python36\Lib\distutils\dist.py", line 955, in run_commands
        self.run_command(cmd)
      File "c:\python36\Lib\distutils\dist.py", line 974, in run_command
        cmd_obj.run()
      File "C:\Users\user\AppData\Local\Temp\pip-install-p9rgnm5n\pulumi-aws\setup.py", line 11, in run
        check_call(['pulumi', 'plugin', 'install', 'resource', 'aws', 'v0.18.11'])
      File "c:\python36\Lib\subprocess.py", line 291, in check_call
        raise CalledProcessError(retcode, cmd)
    subprocess.CalledProcessError: Command '['pulumi', 'plugin', 'install', 'resource', 'aws', 'v0.18.11']' returned non-zero exit status 4294967295.
    ----------------------------------------
ERROR: Command "'c:\users\user\dropbox\pocs\quickstart\venv\scripts\python.exe' -u -c 'import setuptools, tokenize;__file__='"'"'C:\\Users\\user\\AppData\\Local\\Temp\\pip-install-p9r
gnm5n\\pulumi-aws\\setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"
'"'))' install --record 'C:\Users\user\AppData\Local\Temp\pip-record-4fgcondd\install-record.txt' --single-version-externally-managed --compile --install-headers 'c:\users\user\dropbox\pocs\quickstart\venv\include\site\python3.6\pulumi-aws'" failed with error code 1 in C:\Users\user\AppData\Local\Temp\pip-install-p9rgnm5n\pulumi-aws\
```

**Comment by mikhailshilkov** [comment]
I spent some time trying to hunt this error. Everything below applies to my laptop running the latest Windows 10.

Here are some of my findings:

1. `Access is denied` comes from `os.Rename` function call in `workspace/plugins/Install` function during the rename of a temporary folder to the plugin folder. `Access is denied` error is returned after a long pause (~80 seconds, but the value somewhat varies).

1. Sometimes, the call succeeds immediately, but most often it fails. This slows down the trial&error, but overall I can repro quite easily.

1. It  does not seem to be a matter of open file handles:

    - Sysinternals `handles` tool reports `No matching handles found.` during the stuck rename.
    - If I remove `Rename` call, quit `pulumi`, and then try to rename the folder in Windows, I get a similar pause until the rename finally succeeds

1. I disabled all anti-virus / defender / indexing software that I could find on my laptop with no visible effect.

1. Changing the operation of renaming the folder to the renaming of the downloaded file leads to the same pause on file rename.

1. Somewhat surprisingly, changing the code to do the buffered write:

    ``` go
    case tar.TypeReg:
        data, _ := ioutil.ReadAll(r)
        ioutil.WriteFile(path, data, os.FileMode(header.Mode))
    ```
    seems to fix the problem! We talk 100+ MB of buffer, so I'm not sure we want to go this route.

1. After the `Access is denied` error is returned, the folder is actually "unlocked", so the subsequent retry of renaming seems to always immediately succeed. That's the second direction for a workaround that I found so far: just retry the first failure:

    ``` go
    if err := os.Rename(tempDir, finalDir); err != nil && !os.IsExist(err) {
        if err := os.Rename(tempDir, finalDir); err != nil && !os.IsExist(err) {
            return errors.Wrap(err, "moving plugin")
        }
    }
    ```

    The plugin installation is still quite slow with no feedback from CLI at the renaming stage, but it eventually succeeds.

Let me know if you have other ideas to try, or you want me to open a PR for a workaround above.

**Comment by lukehoban** [comment]
> or you want me to open a PR for a workaround above

If you have a fix you believe will work here - would love a PR!

**Comment by lukehoban** [comment]
I believe we believe that #2695 fixed this.  So I'll close this out.  We should reopen if we continue to see occurrences of the issue.

**Comment by lukehoban** [comment]
@pandleigh Can you share exactly what you see?  This was closed because we believed it was fixed with #3008, and have not had a repro for it after that change - but if there are still cases where this can be triggered - we can definitely re-open.

**Comment by pandleigh** [comment]
we use a file encryption software vera on all machines in our company, that software seemed to have locked my folder C:\users\[My Windows Login]\.pulumi folder .. I ran "vera" manually and now it is working. Is there a way to use a different location for ".pulumi" folder in case this happens again

**Comment by pandleigh** [comment]
@lukehoban  here is what i see... i run "pulumi up" for an aws-csharp project. I get an error "error: no resource plugin 'aws-v2.13.0' found in the workspace or on your $PATH, install the plugin using `pulumi plugin install resource aws v2.13.0`"...

so i ran "pulumi plugin install resource aws v2.13.0"  which returns "Moving plugin...error: installing [resource plugin aws-2.13.0] from : moving plugin: rename C:\Users\parag.andleigh\.pulumi\plugins\resource-aws-v2.13.0.tmp163307875 C:\Users\parag.andleigh\.pulumi\plugins\resource-aws-v2.13.0: Access is denied" and i am running it as an "Administrator" in a powershell window

**Comment by pandleigh** [comment]
@lukehoban right now nothing works for me. is there a way to download that plugin separately and copy it manually?

**Comment by serg4kostiuk** [comment]
@lukehoban the same for me:
`C:\Windows\system32>pulumi plugin install resource aws v2.13.0
[resource plugin aws-2.13.0] installing
Downloading plugin: 66.55 MiB / 66.55 MiB [=========================] 100.00% 4s
Moving plugin...error: installing [resource plugin aws-2.13.0] from : moving plugin: rename C:\Users\myuser\.pulumi\plugins\resource-aws-v2.13.0.tmp648263483 C:\Users\myuser\.pulumi\plugins\resource-aws-v2.13.0: Access is denied.`
`C:\Windows\system32>python -V
Python 3.8.4

C:\Windows\system32>pulumi version
v2.6.1`

**Comment by leezen** [comment]
@pandleigh discussed this in Community Slack and observed the machine was running some kind of filesystem protection/encryption system that was preventing it from doing things within the home directory. It seemed like changing the `PULUMI_HOME` environment variable helped.

@serg4kostiuk Curious if you have a similar situation or if you have more details more about the symptoms you're seeing? Are you able to manually create/delete/rename files in that directory yourself?

**Comment by serg4kostiuk** [comment]
Hi @leezen!
I've tried your solution with  `PULUMI_HOME ` environment variable and it doesn't work for me as well. I cannot chose another directory\disk except `C` disk in system because I have only one.
Changed  `PULUMI_HOME ` variable raised the same error:
`C:\Users\myuser\Documents\git\aws-pulumi-python
λ pulumi plugin install resource aws v2.13.0
[resource plugin aws-2.13.0] installing
Downloading plugin: 66.55 MiB / 66.55 MiB [=========================] 100.00% 6s
Moving plugin...error: installing [resource plugin aws-2.13.0] from : moving plugin: rename C:\Users\myuser\Documents\pulumi\plugins\resource-aws-v2.13.0.tmp592995983 C:\Users\myuser\Documents\pulumi\plugins\resource-aws-v2.13.0: Access is denied.`
I can manually create\rename\delete folders and files in the destination folder.

**Comment by leezen** [comment]
@serg4kostiuk Thanks for that. I'll re-open this issue. If you could please provide more details about your OS and the execution context (e.g. is this via cmd.exe, powershell, something else), that would be super helpful.

**Comment by MatthewHarrisZA** [comment]
I also get this on Windows 10 with the Azure plugin. Essentially:

resource-azure-v3.13.0.tmp873795903 folder cannot be renamed to resource-azure-v3.13.0

When I manually rename the folder I can then run "pulumi up"

**Comment by devopsorion** [comment]
@serg4kostiuk I'm having this issues and I change PULUMI_HOME to a directory that I know is excluded by McAfee.

**Comment by dbeattie71** [comment]
@serg4kostiuk  I'm having this issues and I change PULUMI_HOME to a directory that I know is excluded by McAfee.

**Comment by dbeattie71** [comment]
This is strange,
https://stackoverflow.com/questions/56075774/golang-os-renamefromdir-todir-not-working-in-windows

I closed windows file explorer and the IDE's I had open and installing the plugin worked.

**Comment by justinvp** [comment]
It looks like golang users are running into the same problem on Windows when running `go build`, `go get`, etc., when module zips are downloaded and extracted to a temp dir and then renamed with `os.Rename` to the final dir. https://github.com/golang/go/issues/36568. Their solution is to extract the zip in-place in the final destination, rather than extracting to a temp dir and calling `os.Rename`.

**Comment by MisinformedDNA** [comment]
I'm running into this problem on Windows 10 v1909 using both Powershell and Command line, non-admin and admin. McAfee is also running (and I can't disable it).

Trying again doesn't work well for me. It's worked a handful of times, but not reliably. Here's what I saw just now when trying again on Powershell in Admin mode:

![image](https://user-images.githubusercontent.com/1784452/95483249-9ebcac80-0954-11eb-9462-4634d64d08f5.png)

One time, I was able to go the plugins folder and rename the temp directory, but normally the directory gets wiped out right away.

I did find an interesting workaround: When the files are finished downloading, but before the rename occurs, you can select the text to stop the processing.

Now, go to the plugins directory and make a copy of the downloaded files:

![image](https://user-images.githubusercontent.com/1784452/95483847-5b167280-0955-11eb-9a4e-c2430c0272d7.png)

Go back to the Powershell window and right-click to unselect the text. Now the command completes successfully.

![image](https://user-images.githubusercontent.com/1784452/95484121-b183b100-0955-11eb-914f-0bf5c413359b.png)

![image](https://user-images.githubusercontent.com/1784452/95484778-8483ce00-0956-11eb-9832-c28f70b6314d.png)

**Comment by dbeattie71** [comment]
@MisinformedDNA In some cases, I have to close file explorer.

**Comment by MisinformedDNA** [comment]
@dbeattie71 Fails with file explorer closed as well

**Comment by dbeattie71** [comment]
@MisinformedDNA Did you try changing PULUMI_HOME ? I switched it to a a directory I knew was excluded from malware junk running.

**Comment by MisinformedDNA** [comment]
I don't think such a folder exists for me.

**Comment by MisinformedDNA** [comment]
Here's the logs with the [latest verbosity updates](https://github.com/pulumi/pulumi/pull/5549):

> C:\Windows\System32> pulumi plugin install resource azure v3.20 -v=9 --logtostderr
> I1019 09:18:25.309573   25924 sink.go:146] defaultSink::Infoerr([resource plugin azure-3.20.0] installing)
> [resource plugin azure-3.20.0] installing
> I1019 09:18:25.320580   25924 plugins.go:591] SelectCompatiblePlugin(..., azure): beginning
> I1019 09:18:25.320580   25924 plugins.go:624] SelectCompatiblePlugin(..., azure): skipping plugin azure-3.14.0: semver mismatch
> I1019 09:18:25.320580   25924 plugins.go:624] SelectCompatiblePlugin(..., azure): skipping plugin azure-3.15.0: semver mismatch
> I1019 09:18:25.320580   25924 plugins.go:624] SelectCompatiblePlugin(..., azure): skipping plugin azure-3.22.0: semver mismatch
> I1019 09:18:25.321596   25924 plugins.go:624] SelectCompatiblePlugin(..., azure): skipping plugin azure-3.23.0: semver mismatch
> I1019 09:18:25.321596   25924 plugins.go:630] SelectCompatiblePlugin(..., azure): failed to find match
> I1019 09:18:25.321596   25924 plugins.go:201] azure downloading from https://get.pulumi.com/releases/plugins
> I1019 09:18:25.321596   25924 plugins.go:208] full plugin download url: https://get.pulumi.com/releases/plugins/pulumi-resource-azure-v3.20.0-windows-amd64.tar.gz
> I1019 09:18:25.321596   25924 plugins.go:218] plugin install request headers: map[User-Agent:[pulumi-cli/1 (; windows)]]
> I1019 09:18:26.050154   25924 plugins.go:225] plugin install response headers: map[Accept-Ranges:[bytes] Age:[718] Content-Disposition:[attachment; filename=pulumi-resource-azure-v3.20.0-windows-amd64.tar.gz] Content-Length:[56672653] Content-Type:[application/x-gzip] Date:[Mon, 19 Oct 2020 14:06:28 GMT] Etag:["15c67f924f11da53ee26c90da2405c83-11"] Last-Modified:[Fri, 04 Sep 2020 20:35:48 GMT] Server:[AmazonS3] Via:[1.1 548c2d28df76d61dde347634b8c93b73.cloudfront.net (CloudFront)] X-Amz-Cf-Id:[nd7WpLnvrNp9EnZIop1SmMQ3Ib986lBoIYPwvDBztbuL9qsbmejA-g==] X-Amz-Cf-Pop:[LAX50-C1] X-Cache:[Hit from cloudfront]]
> Downloading plugin: 0 B / 54.05 MiB [----------------------------------]   0.00%I1019 09:18:26.053156   25924 plugin_install.go:135] [resource plugin azure-3.20.0] installing tarball ...
> Downloading plugin: 54.05 MiB / 54.05 MiB [========================] 100.00% 19s
> Moving plugin...I1019 09:18:45.718302   25924 plugins.go:304] moving plugin from "C:\\Users\\50587\\.pulumi\\plugins\\resource-azure-v3.20.0.tmp293822791" to "C:\\Users\\50587\\.pulumi\\plugins\\resource-azure-v3.20.0"
> I1019 09:18:46.971051   25924 sink.go:154] defaultSink::Error(error: installing [resource plugin azure-3.20.0] from : moving plugin: rename C:\Users\50587\.pulumi\plugins\resource-azure-v3.20.0.tmp293822791 C:\Users\50587\.pulumi\plugins\resource-azure-v3.20.0: Access is denied.
> main.newPluginInstallCmd.func1
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/pkg/cmd/pulumi/plugin_install.go:137
> github.com/pulumi/pulumi/sdk/v2/go/common/util/cmdutil.RunFunc.func1
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/sdk/go/common/util/cmdutil/exit.go:96
> github.com/pulumi/pulumi/sdk/v2/go/common/util/cmdutil.RunResultFunc.func1
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/sdk/go/common/util/cmdutil/exit.go:112
> github.com/spf13/cobra.(*Command).execute
>         D:/a/pulumi/pulumi/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846
> github.com/spf13/cobra.(*Command).ExecuteC
>         D:/a/pulumi/pulumi/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
> github.com/spf13/cobra.(*Command).Execute
>         D:/a/pulumi/pulumi/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
> main.main
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/pkg/cmd/pulumi/main.go:48
> runtime.main
>         C:/hostedtoolcache/windows/go/1.15.2/x64/src/runtime/proc.go:204
> runtime.goexit
>         C:/hostedtoolcache/windows/go/1.15.2/x64/src/runtime/asm_amd64.s:1374
> )
> error: installing [resource plugin azure-3.20.0] from : moving plugin: rename C:\Users\50587\.pulumi\plugins\resource-azure-v3.20.0.tmp293822791 C:\Users\50587\.pulumi\plugins\resource-azure-v3.20.0: Access is denied.
> main.newPluginInstallCmd.func1
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/pkg/cmd/pulumi/plugin_install.go:137
> github.com/pulumi/pulumi/sdk/v2/go/common/util/cmdutil.RunFunc.func1
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/sdk/go/common/util/cmdutil/exit.go:96
> github.com/pulumi/pulumi/sdk/v2/go/common/util/cmdutil.RunResultFunc.func1
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/sdk/go/common/util/cmdutil/exit.go:112
> github.com/spf13/cobra.(*Command).execute
>         D:/a/pulumi/pulumi/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846
> github.com/spf13/cobra.(*Command).ExecuteC
>         D:/a/pulumi/pulumi/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
> github.com/spf13/cobra.(*Command).Execute
>         D:/a/pulumi/pulumi/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
> main.main
>         D:/a/pulumi/pulumi/src/github.com/pulumi/pulumi/pkg/cmd/pulumi/main.go:48
> runtime.main
>         C:/hostedtoolcache/windows/go/1.15.2/x64/src/runtime/proc.go:204
> runtime.goexit
>         C:/hostedtoolcache/windows/go/1.15.2/x64/src/runtime/asm_amd64.s:1374

**Comment by dbeattie71** [comment]
I can say that since moving the .pulumi directory to a different directory that has exclusions for the virus and malware stuff, I haven't had problems.  On my home machine, I can leave the .pulumi directory under users but on my work machine I have to move it.

**Comment by justinvp** [comment]
@dbeattie71, this should now be addressed as of Pulumi CLI v2.14.0 released Nov 18th. Let me know if you're still experiencing any issues on v2.14.0+.
